### PR TITLE
Fix NPE within annotation attachments

### LINKED ISSUE #41038: [Bug]: Getting NPE when passing an instance method to an annotation
### Description

Following code results in a bad sad error:

```ballerina
import ballerina/io;
public type ResourceConfigRec record {|
    function prefetch;
|};

# The annotation to configure the load method with batch functions.
public annotation ResourceConfigRec ResourceConfig on object function;

isolated distinct service class Author {

    @ResourceConfig {
        prefetch: self.prefetchBooks
    }
    isolated resource function get books() returns string {
        return "";
    }

    public function prefetchBooks() {
        io:println("Prefetching books");
    }
}
``` 

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by mindula**
## Purpose
$tilte

Fixes #41038

## Samples
Screenshot 2024-05-16 at 14 46 20

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 1]
```diff
@@ -4683,10 +4683,12 @@ private void validateAnnotationAttachmentExpr(BLangAnnotationAttachment annAttac
         }
 
         if (annAttachmentNode.expr != null) {
+            data.commonAnalyzerData.withinAnnotationExpr = true;
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:4686]:
```suggestion
            boolean prevWithinAnnotationExpr = data.commonAnalyzerData.withinAnnotationExpr;           data.commonAnalyzerData.withinAnnotationExpr = true;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -4683,10 +4683,12 @@ private void validateAnnotationAttachmentExpr(BLangAnnotationAttachment annAttac
...
             this.typeChecker.checkExpr(annAttachmentNode.expr, data.env,
                     referredAnnotType.tag ==
                             TypeTags.ARRAY ? ((BArrayType) referredAnnotType).eType : annotType, data.prevEnvs,
                     data.commonAnalyzerData);
+            data.commonAnalyzerData.withinAnnotationExpr = false;
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:4691]: Use the previous value of the `withinAnnotationExpr` in here.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -3182,6 +3182,9 @@ public void visit(BLangSimpleVarRef varRefExpr, AnalyzerData data) {
             // TODO: call to isInLocallyDefinedRecord() is a temporary fix done to disallow local var references in
             //  locally defined record type defs. This check should be removed once local var referencing is supported.
             if (((symbol.tag & SymTag.VARIABLE) == SymTag.VARIABLE)) {
+                if (data.commonAnalyzerData.withinAnnotationExpr && varName.getValue().equals("self")) {
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3185]: use the constant value defined for `self` keyword.

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/within_annotation_attachment_negative.bal [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    @ResourceConfig {
+        prefetch: self.prefetchBooks
+    }
+    isolated resource function get books() returns string {
+        return "";
+    }
+
+    public function prefetchBooks() {
+    }
+}
```
Comment (LakshanWeerasinghe) [inline tests/jballerina-unit-test/src/test/resources/test-src/annotations/within_annotation_attachment_negative.bal:34]: Add tests for objects and service type definitions as well.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 3 above)
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3185]: IMO, this is not correct. Before reaching this logic, we check for `self` within the main scope and set the symbol for self, which means we are allowing the self variable within that scope. This is not correct. At line 3171(`lookupMainSpaceSymbolInPackage`), when checking for the symbol within the scope, we should not see self for this annotation scope. 
What we did here is set the symbol of the variable and then handle it as a special case.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 5]
```diff
@@ -7169,6 +7169,7 @@ public static class CommonAnalyzerData {
         boolean breakToParallelQueryEnv = false;
         int letCount = 0;
         boolean nonErrorLoggingCheck = false;
+        boolean withinAnnotationExpr = false;
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:7172]: If we do not define `self` within the annotation scope, there is no need to define a special variable for it.

**Comment by KavinduZoysa** [comment]
As per https://github.com/ballerina-platform/ballerina-lang/pull/42762#discussion_r1615099103, the implementation is not correct. Closing the PR until required changes are made.

**Pull request closed without merging**